simSPI/linear_simulator/params_utils.py

```diff
@@ -0,0 +1,352 @@
+        ----------
+        config: class
+
+        Returns
+        -------
+        params_generator: class
+        """
+        if config.starfile_available:
+            return starfile_params(config)
+        else:
+            return distributional_params(config)
+
+
+class Iparams(metaclass=ABCMeta):
+    """Abstract class for the params generator factory with get_params method."""
+
+    @abstractstaticmethod
+    def get_params(self):
+        """Get the parameters."""
+
+
+"""Module to generate parameters using the input star file."""
+
+
+class starfile_params(Iparams):
+    """Class to generate parameters using the input star file.
+
+    Parameters
+    ----------
+    config: class
+        Class containing parameters of the dataset generator.
+
+    """
+
+    def __init__(self, config):
+
+        self.counter = 0
+        self.invert_hand = False
+        self.df = starfile.read(config.input_starfile_path)
+        check_star_file(config.input_starfile_path)
+        self.config = config
+        print("Reading parameters from the input starfile.")
+
+    def get_params(self):
+        """Get the parameters from the starfile.
+
+        Returns
+        -------
+        rot_params: dict of type str to {tensor}
+            Dictionary of rotation parameters for a projection chunk
+        ctf_params: dict of type str to {tensor}
+            Dictionary of Contrast Transfer Function (CTF) parameters
+            for a projection chunk
+        shift_params: dict of type str to {tensor}
+            Dictionary of shift parameters for a projection chunk
+        """
+        self.particle = self.df["particles"].iloc[
+            self.counter : self.counter + self.config.chunks
+        ]
+        self.counter += self.config.chunks
+        return self.get_rotmat(), self.get_ctf_params(), self.get_shift_params()
+
+    def get_rotmat(self):
+        """Get the parameters for the rotation of the projections from starfile.
+
+        Returns
+        -------
+        rotmat_params: dict of type str to {tensor}
+            dictionary containing:
+            "rotmat": torch.Tensor
+                Tensor of size (chunks,3,3) that containing the rotation matrix for
+                each projection
+            "relion_AnglePsi": torch.Tensor
+                Tensor of size (chunks,) that contains the Psi angle (in degrees)
+                 for ZYZ rotation
+            "relion_AngleTilt": torch.Tensor
+                Tensor of size (chunks,) that contains the Tilt angle (in degrees)
+                 for ZYZ rotation
+            "relion_AngleRot": torch.Tensor
+                Tensor of size (chunks,) that contains the Rot angle (in degrees)
+                for ZYZ rotation
+        """
+        AnglePsi = np.array(self.particle["rlnAnglePsi"], ndmin=2).squeeze()
+        AngleTilt = np.array(self.particle["rlnAngleTilt"], ndmin=2).squeeze()
+        AngleRot = np.array(self.particle["rlnAngleRot"], ndmin=2).squeeze()
```

Comment:
I would modularize things so that we read it a compSPI (non-relion) star file, and there's a separate text to text .star file conversion step.

You can use the same headings if you want, but just rename `rlnAnglePsi` to `AnglePsi` or something